src/mysql_protocol.erl

```diff
@@ -1630,45 +1654,65 @@ server_info_test() ->
 decode_text_test() ->
     %% Int types
     lists:foreach(fun (T) ->
-                      ?assertEqual(1, decode_text(#col{type = T}, <<"1">>))
+                      ?assertEqual(1, decode_text(#col{type = T}, #decode_options{}, <<"1">>))
                   end,
                   [?TYPE_TINY, ?TYPE_SHORT, ?TYPE_LONG, ?TYPE_LONGLONG,
                    ?TYPE_INT24, ?TYPE_YEAR]),
 
     %% BIT
-    <<217>> = decode_text(#col{type = ?TYPE_BIT, length = 8}, <<217>>),
+    <<217>> = decode_text(#col{type = ?TYPE_BIT, length = 8}, #decode_options{}, <<217>>),
 
     %% Floating point and decimal numbers
     lists:foreach(fun (T) ->
-                      ?assertEqual(3.0, decode_text(#col{type = T}, <<"3.0">>))
+                      ?assertEqual(3.0, decode_text(#col{type = T}, #decode_options{}, <<"3.0">>))
                   end,
                   [?TYPE_FLOAT, ?TYPE_DOUBLE]),
     %% Decimal types
     lists:foreach(fun (T) ->
-                      ColDef = #col{type = T, decimals = 1, length = 4},
-                      ?assertMatch(3.0, decode_text(ColDef, <<"3.0">>))
+                      ColDef = #col{type = T},
+                      ?assertMatch(3.0, decode_text(ColDef#col{decimals = 1, length = 4}, #decode_options{}, <<"3.0">>)),
+                      %% Decimal Decode Options
+                      %% A small value like this would be returned as a float
+                      %% if decode_decimal=auto. We want to test forcing the
+                      %% binary return
+                      ?assertMatch(<<"3.0">>, decode_text(ColDef#col{decimals = 1, length = 4},
+                                                          #decode_options{decode_decimal=binary}, <<"3.0">>)),
+
+                      %% We don't check for accuracy because this number is
+                      %% large enough that the value will lose its precision
+                      %% when converting to a float, so we just check that the
+                      %% return value is a float and move on
+                      ?assert(is_float(decode_text(ColDef#col{decimals = 12, length = 23},
+                                                   #decode_options{decode_decimal=number}, <<"123456789.456789123456789">>))),
```

Comment:
I think we should check the value. Even if a float is not exact, both the parsing and the float we compare against will be rounded in the same way:

```
1> 123456789.456789123456789.
123456789.45678912
2> 123456789.456789123456789 =:= 123456789.456789123456789.
true
3> 123456789.456789123456789 =:= 123456789.45678912.       
true
```

Therefore, I think we can safely match like this:

```suggestion
                      %% When decode_decimal=number, we expect a float and accept the
                      %% precision loss.
                      ?assertMatch(123456789.456789123456789,
                                   decode_text(ColDef#col{decimals = 12, length = 23},
                                               #decode_options{decode_decimal=number},
                                               <<"123456789.456789123456789">>)),
```

Reply:
Acknowledged.